File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n, m;
  cin >> n >> m;

  vector<int> v(100 + 1, 0);
  vector<int> visited(100 + 1, 0);
  queue<tuple<int, int>> q;

  for (int i = 0; i < n + m; ++i) {
    int x;
    cin >> x;
    cin >> v[x];
  }

  q.push({1, 0});
  int result = 0;

  while (!q.empty()) {
    auto [now_num, now_step] = q.front();
    q.pop();

    if (now_num == 100) {
      result = now_step;
      break;
    }

    for (int i = 1; i <= 6; ++i) {
      int new_num = now_num + i;
      if (new_num > 100) {
        continue;
      }

      if (v[new_num] != 0) {
        new_num = v[new_num];
      }

      if ((visited[new_num] != 0 && visited[new_num] <= visited[now_num])) {
        continue;
      }

      visited[new_num] = now_step + 1;
      q.push({new_num, visited[new_num]});
    }
  }

  cout << result;

  return 0;
}

File: 백준/Gold/16928. 뱀과 사다리 게임/뱀과 사다리 게임.py
from collections import deque
import sys


def main():
  readline = sys.stdin.readline

  max_n = 100
  v = [0] * (max_n + 1)
  visited = [0] * (max_n + 1)

  n, m = map(int, readline().split())

  for _ in range(n + m):
    x, y = map(int, readline().split())
    v[x] = y

  dq = deque()
  dq.append((1, 0))
  result = 0

  while dq:
    now_num, now_step = dq.popleft()

    if now_num == max_n:
      result = now_step
      break

    for i in range(1, 7):
      new_num = now_num + i

      if new_num > max_n:
        continue

      if v[new_num] != 0:
        new_num = v[new_num]

      if visited[new_num] != 0 and visited[new_num] <= now_step + 1:
        continue

      visited[new_num] = now_step + 1
      dq.append((new_num, now_step + 1))

  print(result)


if __name__ == "__main__":
  main()
